cmd/f3/observer.go

```diff
@@ -0,0 +1,320 @@
+			return fmt.Errorf("initialzing pubsub: %w", err)
+		}
+
+		manifestServerID, err := peer.Decode(c.String("manifestID"))
+		if err != nil {
+			return fmt.Errorf("decoding manifestServerID: %w", err)
+		}
+		manifestProvider := manifest.NewDynamicManifestProvider(&manifest.Manifest{}, ds, pubSub, manifestServerID)
+		if err != nil {
+			return fmt.Errorf("initialzing manifest sender: %w", err)
+		}
+		err = manifestProvider.Start(c.Context)
+		if err != nil {
+			return fmt.Errorf("starting manifest provider: %w", err)
+		}
+
+		go func() {
+			ticker := time.NewTicker(15 * time.Second)
+			defer ticker.Stop()
+			for c.Context.Err() == nil {
+				select {
+				case <-ticker.C:
+					n := len(host.Network().Peers())
+					if n < 5 {
+						log.Infow("low peers, bootstrapping again", "NPeers", n)
+						connectToBootstrappers()
+					}
+				case <-c.Context.Done():
+				}
+			}
+		}()
+
+		var closer func()
+		var runningNetwork gpbft.NetworkName
+		for c.Context.Err() == nil {
+			select {
+			case manif := <-manifestProvider.ManifestUpdates():
+				log.Infof("Got manifest for network %s\n", manif.NetworkName)
+				if runningNetwork == manif.NetworkName {
+					log.Infof("same network, continuing")
+					continue
+				}
+				if closer != nil {
+					closer()
+				}
+
+				networkCtx, c := context.WithCancel(c.Context)
+				_ = networkCtx
+				closer = c
+				err := observeManifest(networkCtx, manif, pubSub)
+				if err != nil {
+					log.Errorw("starting manifest observer", "error", err, "networkName", manif.NetworkName)
+				}
+			case <-c.Context.Done():
+			}
+		}
+		return nil
+	},
+}
+
+func observeManifest(ctx context.Context, manif *manifest.Manifest, pubSub *pubsub.PubSub) error {
+	dir := msgdump.DirForNetwork("./observer", manif.NetworkName)
+	parquetWriter, err := msgdump.NewParquetWriter[msgdump.ParquetEnvelope](dir)
+	if err != nil {
+		return fmt.Errorf("creating parquet file: %w", err)
+	}
+	logFileName := filepath.Join(dir, "msgs.ndjson")
+	msgLogFile, err := os.OpenFile(logFileName, os.O_WRONLY|os.O_CREATE|os.O_CREATE, 0660)
+	if err != nil {
+		return fmt.Errorf("creating msg log file: %w", err)
+	}
+	jsonWriter := json.NewEncoder(msgLogFile)
+
+	err = pubSub.RegisterTopicValidator(manif.PubSubTopic(), func(_ context.Context, _ peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
+		var gmsg gpbft.GMessage
+		if err := gmsg.UnmarshalCBOR(bytes.NewReader(msg.Data)); err != nil {
+			return pubsub.ValidationReject
+		}
+		msg.ValidatorData = gmsg
+
+		return pubsub.ValidationAccept
+	})
+	if err != nil {
+		return fmt.Errorf("registering topic validator: %w", err)
+	}
+
+	topic, err := pubSub.Join(manif.PubSubTopic(), pubsub.WithTopicMessageIdFn(pubsub.DefaultMsgIdFn))
+	if err != nil {
+		return fmt.Errorf("joining topic: %w", err)
+	}
+
+	sub, err := topic.Subscribe()
+	if err != nil {
+		return fmt.Errorf("subscribing to topic: %w", err)
+	}
+
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, syscall.SIGUSR1)
```

Comment:
might as well handle interrupts as well? Or will the context do that?

Reply:
Context will do that